ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala

```diff
@@ -0,0 +1,122 @@
+    implicit grpcExecutionContext: ExecutionContext,
+    logCtx: LoggingContext)
+    extends ParticipantPruningServiceGrpc.ParticipantPruningService
+    with GrpcApiService {
+
+  private val logger = ContextualizedLogger.get(this.getClass)
+
+  override def bindService(): ServerServiceDefinition =
+    ParticipantPruningServiceGrpc.bindService(this, DirectExecutionContext)
+
+  override def prune(request: PruneRequest): Future[Empty] = {
+    val submissionId =
+      if (request.submissionId.isEmpty)
+        SubmissionId.assertFromString(UUID.randomUUID().toString)
+      else
+        SubmissionId.assertFromString(request.submissionId)
+
+    LoggingContext.withEnrichedLoggingContext("submissionId" -> submissionId) { implicit logCtx =>
+      val pruneUpToOffset: Either[StatusRuntimeException, (Offset, String)] = for {
+        pruneUpToProto <- request.pruneUpTo.toRight(
+          ErrorFactories.invalidArgument("prune_up_to not specified"))
+
+        pruneUpToString <- pruneUpToProto.value match {
+          case LedgerOffset.Value.Absolute(offset) => Right(offset)
+          case LedgerOffset.Value.Boundary(boundary) =>
+            Left(
+              ErrorFactories.invalidArgument(
+                s"prune_up_to needs to be absolute and not a boundary ${boundary.toString}"))
+          case LedgerOffset.Value.Empty =>
+            Left(
+              ErrorFactories.invalidArgument(
+                s"prune_up_to needs to be absolute and not empty ${pruneUpToProto.value}"))
+        }
+
+        pruneUpTo <- Ref.HexString
+          .fromString(pruneUpToString)
+          .left
+          .map(err =>
+            ErrorFactories.invalidArgument(
+              s"prune_up_to needs to be a hexadecimal string and not ${pruneUpToString}: ${err}"))
+          .map(Offset.fromHexString)
+
+      } yield (pruneUpTo, pruneUpToString)
+
+      pruneUpToOffset.fold(
+        Future.failed, {
+          case (pruneUpTo, pruneUpToString) =>
+            (for {
+              ledgerEnd <- readBackend.currentLedgerEnd()
+              _ <- if (pruneUpToString < ledgerEnd.value) Future.successful(())
+              else
+                Future.failed(
+                  ErrorFactories.invalidArgument(
+                    s"prune_up_to needs to be before ledger end ${ledgerEnd.value}"))
+
+              _ <- FutureConverters.toScala(writeBackend.prune(pruneUpTo, submissionId)).flatMap {
+                case NotPruned(status) => Future.failed(ErrorFactories.fromGrpcStatus(status))
+                case ParticipantPruned =>
+                  logger.info(s"Pruned participant ledger up to ${pruneUpTo.toString} inclusively.")
+                  Future.successful(())
+              }
+
+              pruneResponse <- readBackend
+                .prune(pruneUpTo)
+                .map { _ =>
+                  logger.info(s"Pruned ledger api server index up to ${pruneUpTo} inclusively.")
+                  Empty()
+                }
```

Comment:
This is also a good candidate to be put into its own method.

Ideally you would end up with something like:

```scala
val pruneUpTo =
  for {
    _ <- checkPruneUpTo(request)
    pruneUpTo <- decodePruneUpTo(request)
    _ <- validatePruneUpTo(pruneUpTo)
  } yield prunedUpTo

prunedUpTo.fold(Future.failed, doPrune(writeBackend, readBackend, submissionId, _))
```

Or something like this.

Reply:
Okay, I took a stab at putting each of the steps (and to the note above: `validateRequest` as a wrapper function for validation) after trying out a few arrangements, I think I ended up with the simplest, most readable one by keeping the submissionId at the outermost level, so that I only end up with one `LoggingContext` wrapper.